### PR TITLE
feat(groups): group→role mappings with managed-role reconciliation (#1204)

### LINKED ISSUE #1204: Group sync Phase 1: group→role mappings with managed-role reconciliation (sync + login)
Part of Epic #1202 (Phase 1). Groups drive roles automatically: admin-managed mappings from synced Google groups onto AI Studio roles, reconciled by the hourly sync and at login. Roles remain the permission backbone, so capabilities, API scopes, and agent-skill checks keep working with zero query changes.

## Scope

- **Migration**: `group_role_mappings` (group_email lowercase, role_id FK CASCADE, unique pair, timestamps). Identifier is group **email** (rename-safe), matching Phase 0 tables.
- **Migration**: add `source varchar` (`'manual'` default | `'group-sync'`) to `user_roles` — the managed-role flag.
- **Reconciliation** (one shared module, called from both the sync Lambda path and login):
  - computed = roles mapped from the user's current group memberships
  - add missing computed roles with `source='group-sync'`
  - remove only `source='group-sync'` rows no longer computed — **admin-assigned (manual) roles are never touched**
  - bump `users.role_version` only when membership actually changed
- **Login-time hook**: run reconciliation for the signing-in user in `getCurrentUserAction()` (canonical provisioning path) with parity in `lib/auth/resolve-user.ts`. Keep the numeric-username→student / else→staff default as fallback for users in zero mapped groups (retirement tracked in Phase 4).
- **Admin UI**: mappings CRUD on `/admin/groups` (group picker from synced groups, role picker from live `getRoles()`).
- **While in this area**: `components/user/multi-role-selector.tsx` and `components/user/user-role-form.tsx` hardcode the 3 role names in JSX — switch to dynamic `getRoles()` so mapped-in roles show correctly.

## Acceptance criteria

- [ ] Adding a mapping grants the role to all current members on next sync/login without manual role edits (verified end-to-end)
- [ ] Removing a mapping (or leaving the group) removes ONLY sync-managed roles; a manually assigned identical role persists (test)
- [ ] `role_version` bumps exactly when roles change (no churn on no-op reconciliation)
- [ ] Role selectors show roles dynamically from the DB
- [ ] Reconciliation targets canonical `lib/db/drizzle/user-roles.ts`; the duplicate `lib/db/user-roles.ts` module either consolidated or explicitly left with a comment (4 importers: atrium requester ×2, delegated-token route, admin users roles route)
- [ ] Lint, typecheck, unit tests for reconciliation edge cases (user in 0 groups, overlapping mappings, manual+managed same role)

## Notes / invariants

- Role grants everywhere match by **name** string (`roles.name`) — mappings store role_id but reconciliation resolves through the existing by-name helpers (learning 2026-06-29-role-grant-name-not-id).
- Capability/scope checks are live DB queries (no caching), so sync-driven role changes take effect on the user's next request — no re-login needed.
- Don't use `undefined` in Drizzle `.set()` for clearable fields (`?? null`).

### PR DESCRIPTION
**Pull request by krishagel**
Closes #1204 (Epic #1202, Phase 1).

## What this ships

Synced Google groups now **drive AI Studio roles**. Admin-managed `group_role_mappings` reconcile into `user_roles` with a managed-role flag, so group membership grants and revokes roles automatically while hand-assigned roles stay untouchable.

- **Migration 109** (registered — renumbered from 108 after #1221 took that slot on dev): `group_role_mappings` (group email → `roles.id`, unique on `(lower(group_email), role_id)`) and `user_roles.source` (`'manual'` default backfills every pre-existing row | `'group-sync'`), CHECK-constrained and mirrored in the Drizzle schema.
- **Reconciliation, two implementations of one documented contract** (the Lambda bundle can't import `@/lib` — same boundary as Phase 0):
  - **Per-user** (`reconcileUserManagedRoles`): runs at session establishment (`getCurrentUserAction`) and JIT provisioning (`resolveUserId` email-link/create paths) — deliberately **not** on the per-request fast path (~13 polling GET routes; three reviewers independently flagged the amplification).
  - **Bulk** (`reconcileManagedRoles`, sync Lambda): set-based pass over all users after each hourly/manual membership sync — temp-table snapshot, `RETURNING`-derived `role_version` bumps.
- **Invariants** (enforced + unit-tested, 16 tests): add missing mapped roles as `group-sync`; remove only `group-sync` rows no longer computed; **manual grants are never touched**; `role_version` bumps only when a row was actually written (verified via `RETURNING`, not the pre-write snapshot — concurrent reconciles can't churn); **the last administrator can never be auto-revoked** (pure `applyLastAdminGuard`, both paths, mirrors the manual-path guard).
- **Admin UI** (`/admin/groups` → Role mappings tab): mapping CRUD with pickers fed by synced groups and live `getRoles()`; the users-page role selectors now read roles dynamically instead of a hardcoded 3-name list.
- **Sync-now status cue** (carried over from #1203 verification feedback): the button shows a spinning "Syncing…" state, polls until the run *settles* (not merely starts — each group commits `last_synced_at` independently), then toasts and auto-refreshes; the timeout branch refreshes too so an all-groups-fail run surfaces its failure counts. "Last run" renders as relative time.
- **Source-aware role editor**: `updateUserRoles` (both the admin route's standalone module and the drizzle copy) no longer full-deletes and re-inserts — managed rows are invisible to it, closing the two silent failure modes (admin revocation that reconciliation instantly undoes; managed roles permanently converted to manual by any admin edit).
- **Propagation**: capability/scope checks are live DB queries, so a reconcile-driven change applies on the user's next request; `role_version` + refresh-session handles the JWT layer — no new mechanism.

## Review round (6 agents) — all confirmed findings fixed in `bc068994`

schema-drift (clean), simplicity, security, adversarial, correctness, SQL. Highlights: **last-administrator lockout guard** (security P2 — an org mapping a group to `administrator` could reach zero admins with no in-app recovery), **email-source divergence** (adversarial P1 — `getCurrentUserAction` reconciled with the never-refreshed `users.email` while JIT paths used the session email, actively flapping roles after a Workspace email change), **source-blind legacy role editor** (adversarial P1 + SQL HIGH), **no-op `role_version` churn under concurrency** (correctness P1), hot-path reconcile removal, premature "Sync complete" on multi-group runs, dead `UserRoleForm` deleted (zero callers; this PR had been making it *more* elaborate), single-CTE mapping upsert.

**Reviewed-and-not-changed, with rationale** (details in the commit message): migration 108's inline CHECK lock (tiny junction table), no `lower(users.email)` index yet (#1207 adds the unique one), bounded double-bump across two racing *true* changes (self-stabilizing; refresh-session compares `!==`).

## Behavior notes for admins

- Removing a group-managed role = remove the user from the Google group or delete the mapping. The role editor intentionally cannot delete managed rows (they'd silently reappear on the next reconcile). A "via group" badge in the role editor is a sensible follow-up if this confuses anyone.
- If the last administrators are all group-managed and the mapped group empties, the guard refuses the revocation and logs `Last-administrator guard tripped` (Lambda error log + per-user warn).

## Verification

- `jest managed-role-reconciliation`: **16/16** (11 diff-matrix + 5 lockout-guard)
- Lambda `bun test`: **14/14**; Lambda `tsc` clean (pinned 5.9.3 via local install — note: bare `bunx tsc` without `bun install` resolves a global TS 6 that errors on this tsconfig; CDK bundling always installs first)
- Full-repo `bun run typecheck`: clean; touched-file lint: clean
- Gated functional E2E `admin-groups.functional.spec.ts` (selection + role-mappings tabs): **2/2 green** on the seeded :3100 harness (screenshots in `.verification/`)
- Pre-push hook: full local E2E suite **217 passed**

## Deploy

Standard full-stack deploy applies migration 109; the sync Lambda picks up the reconciliation pass on its next hourly run. No new settings, no Google-side changes.

https://claude.ai/code/session_012RDk3YXBFfbHh5NS3joUxq

### CODE REVIEW

File: app/(protected)/admin/groups/_components/groups-admin.tsx [hunk 1]
```diff
@@ -73,18 +119,73 @@ export function GroupsAdmin({ initialData, initialError }: GroupsAdminProps) {
...
+      setIsSyncing(false)
+      toast({
+        title: "Sync still running",
+        description: "Taking longer than usual — showing what has finished so far.",
+      })
+      // Refresh here too: covers the all-groups-fail run, where last_synced_at
+      // never advances but the failed-count banner has data worth surfacing.
+      router.refresh()
+    }
   }
```
[the remark this answers is not in the document]
Comment (krishagel) [reply app/(protected)/admin/groups/_components/groups-admin.tsx:206]: Fixed in 961caaf — the sync-now guard now uses a `syncInFlightRef` ref checked synchronously before `setIsSyncing(true)`, closing the same-tick double-click race.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/db/drizzle/user-roles.ts [hunk 2]
```diff
@@ -295,3 +330,274 @@ export async function assignRoleToUser(userId: number, roleId: number) {
...
+            if (adminRole && toRemove.includes(adminRole.id)) {
+              const [remaining] = await tx
+                .select({ count: sql<number>`count(*)::int` })
+                .from(userRoles)
+                .where(
+                  and(
+                    eq(userRoles.roleId, adminRole.id),
+                    sql`${userRoles.userId} <> ${userId}`
+                  )
+                );
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/db/drizzle/user-roles.ts:582]: Fixed in 961caaf — switched to the native `ne` operator.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/db/drizzle/user-roles.ts [hunk 3]
```diff
@@ -96,27 +110,48 @@ export async function updateUserRoles(
...
+            .where(
+              and(
+                eq(userRoles.userId, userId),
+                eq(userRoles.source, "manual"),
+                submittedIds.length > 0
+                  ? sql`${userRoles.roleId} NOT IN (${sql.join(
+                      submittedIds.map((id) => sql`${id}`),
+                      sql`, `
+                    )})`
+                  : sql`TRUE`
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/db/drizzle/user-roles.ts]: Fixed in 961caaf — switched to the native `notInArray` operator.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/db/user-roles.ts [hunk 4]
```diff
@@ -107,30 +107,54 @@ export async function updateUserRoles(userId: number, roleNames: string[]): Prom
...
+          .where(
+            and(
+              eq(userRoles.userId, userId),
+              eq(userRoles.source, 'manual'),
+              submittedIds.length > 0
+                ? sql`${userRoles.roleId} NOT IN (${sql.join(
+                    submittedIds.map(id => sql`${id}`),
+                    sql`, `
+                  )})`
+                : sql`TRUE`
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/db/user-roles.ts]: Fixed in 961caaf — switched to the native `notInArray` operator.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/db/schema/tables/user-roles.ts [hunk 5]
```diff
@@ -3,18 +3,33 @@
...
+ *                only source reconciliation ever adds or removes.
+ */
+export type UserRoleSource = "manual" | "group-sync";
+
 export const userRoles = pgTable("user_roles", {
   id: serial("id").primaryKey(),
   userId: integer("user_id").references(() => users.id, { onDelete: "cascade" }),
   roleId: integer("role_id").references(() => roles.id),
+  // Managed-role flag — see UserRoleSource. Defaults to 'manual' (migration 108).
+  source: varchar("source", { length: 20 }).$type<UserRoleSource>().default("manual").notNull(),
```
[the remark this answers is not in the document]
Comment (krishagel) [reply lib/db/schema/tables/user-roles.ts:25]: Fixed in 961caaf — `updateUser`'s save path is now source-aware (manual rows only, managed rows invisible). Also just deduped this logic: it's now shared across all three write paths via `syncManualUserRoles` in `lib/db/drizzle/user-roles.ts` (7757bff), so a future fix here only needs to land once.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: infra/lambdas/group-sync/db.ts [hunk 6]
```diff
@@ -183,3 +183,132 @@ export async function deactivateGroupsNotIn(
...
+  return sql.begin(async (tx) => {
+    await tx`
+      CREATE TEMP TABLE _computed_roles ON COMMIT DROP AS
+      SELECT DISTINCT u.id AS user_id, grm.role_id
+        FROM group_role_mappings grm
+        JOIN groups g
+          ON lower(g.group_email) = lower(grm.group_email)
+         AND g.is_active = true
+        JOIN group_members gm ON gm.group_id = g.id
+        JOIN users u ON lower(u.email) = lower(gm.member_email)
```
[the remark this answers is not in the document]
Comment (krishagel) [reply infra/lambdas/group-sync/db.ts:237]: Fixed in 961caaf — `getCurrentUserAction` now refreshes `users.email` from the session when it differs (case-insensitive), so the bulk pass has a current join key for any user who has signed in since the rename. Documented residual noted in the commit: between the rename and the next sign-in, one bulk pass may drop managed roles once — self-heals at login.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
🤖 pr-fix routine round 1 complete.

**Already-addressed comments resolved (6):** the 4 gemini-code-assist + 2 chatgpt-codex-connector inline findings were fixed by 961caaf (pushed before this run) — replied on each thread pointing to the fix and marked resolved.

**New fixes pushed (7757bff)**, addressing the two remaining open findings from the claude[bot] review round:
- **Correctness (TOCTOU race):** the last-administrator guard's read-then-delete in both `reconcileUserManagedRoles` (per-user) and the bulk Lambda's `reconcileManagedRoles` (hourly) had no locking — two concurrent revocations could each read the other's uncommitted admin row as "surviving" and both proceed, leaving zero administrators despite the guard. Added a shared `pg_advisory_xact_lock` (same key in both the Next.js app and the Lambda bundle, since the Lambda can't import `@/lib`) around each reconciler's admin-removal branch, following the existing lock pattern in `lib/capabilities/sync.ts`.
- **Simplification (3x duplication):** the source-aware manual-role delete+insert logic was copy-pasted across `actions/admin/user-management.actions.ts#updateUser`, `lib/db/drizzle/user-roles.ts#updateUserRoles`, and `lib/db/user-roles.ts#updateUserRoles` — and had already drifted once (961caaf had to fix the same bug in three places). Extracted to `syncManualUserRoles()` in `lib/db/drizzle/user-roles.ts`; all three call sites now share it.

**Not addressed — deferred, not blocking:** the claude[bot] review's test-coverage suggestion (integration tests for the DB-touching reconcile functions and admin CRUD actions) — the reviewer explicitly framed this as a non-blocking follow-up. Recommend filing a tracking issue if the team wants it before merge.

**Verification note:** this sandboxed routine's network egress does not allow `bun install` (npm registry isn't in the egress allowlist), so I could not run `typecheck`/`lint`/`jest` locally for 7757bff. I did a careful manual review (type consistency with the existing `DbTransaction` signature shared by all three call sites, confirmed the pure functions covered by `tests/unit/managed-role-reconciliation.test.ts` are untouched, checked for other test mocks that might break from the refactor — found none). This repo's CI (Test, Lint, and Type Check) has full network access and will be the actual gate — please keep an eye on it.

- Review comments addressed: 8 (6 already-fixed + 2 new)
- Commits pushed: 1
- Files changed: 4




---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
CI is fully green on 7757bff (Test/Lint/Typecheck, CDK validate, CodeQL, claude-review — all success).

However, the claude-review round triggered by that push (https://github.com/psd401/aistudio/pull/1222#issuecomment-4964318399) surfaced new actionable findings, including one directly on the fix just pushed:

- **Perf/consistency**: the bulk Lambda's advisory lock (`reconcileManagedRoles`) is acquired unconditionally every hourly run, while the per-user path only acquires it when an admin-role removal is actually in play — worth gating the bulk path the same way to avoid unnecessary serialization against the per-user path on every run.
- **Correctness**: `updateUser`'s last-admin guard counts *all* `user_roles` rows for `administrator` (any source) against the *submitted* list, but the write it guards now only ever touches `source='manual'` rows — can false-positive block a legitimate edit when the sole admin holds that role via `group-sync`.
- **Correctness**: optimistic client state in the admin users UI trusts the submitted role list after save, so unchecking a group-managed role appears to succeed in the table even though the DB write silently no-ops.
- **Dead code**: 3 unreachable files (`users-client.tsx`, `users-table.tsx`, `multi-role-selector.tsx`) — candidates for deletion.
- Minor: inactive-group indicator in the mapping picker, one stale comment.

Per this routine's per-fire limit (one PR per run), I'm not starting a new round of fixes in this cycle. Leaving labels as-is so the next scheduled fire (or a human) picks this up as a fresh round.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by krishagel** [comment]
Review loop complete through 4 rounds — all confirmed findings fixed; CI green and mergeable at `9c8538fb`.

**Rounds 3–4 (after the PR-body summary was written):** source-aware last-admin guard in the manual `updateUser` editor (was false-positive-blocking unrelated edits on group-sync admins) + it now takes the shared `pg_advisory_xact_lock` so all THREE admin-revoking paths serialize; `updateUser` returns the actual post-write role set and the users page updates optimistic state from it (unchecking a managed role no longer shows a phantom revocation); per-user reconcile self-throttles (5 min/user/container — `requireRole()` funnels ~40 admin actions through it per call); bulk pass takes the advisory lock only when it actually has an admin deletion candidate; deleted the unreachable `users-client`/`users-table`/`multi-role-selector` trio (~800 lines, nothing imports them — the live page is `users-page-client` → `user-detail-sheet`); widened the load-flaky polling-backoff spec's observation window (failed-all-retries twice under full-suite load, passes in isolation).

**Known follow-ups (non-blocking):** integration tests for the DB-touching reconcilers against local Postgres; a "via group" badge in the role editor; users renamed in Workspace may have managed roles drop once between the rename and their next sign-in (bulk joins on `users.email`, refreshed at session establishment).
